bus-mapping/src/evm/opcodes/callop.rs

```diff
@@ -115,32 +115,52 @@ impl<const N_ARGS: usize> Opcode for CallOpcode<N_ARGS> {
             state.call_context_write(&mut exec_step, call.call_id, field, value);
         }
 
-        state.transfer(
-            &mut exec_step,
-            call.caller_address,
-            call.address,
-            call.value,
-        )?;
+        match call.kind {
+            CallKind::Call => {
+                // Transfer value only for CALL opcode.
+                state.transfer(
+                    &mut exec_step,
+                    call.caller_address,
+                    call.address,
+                    call.value,
+                )?;
+            }
+            CallKind::CallCode => {
+                // For CALLCODE opcode, get caller balance to constrain it should be greater or
+                // equal to stack `value`.
+                let (_, caller_account) = state.sdb.get_account(&call.caller_address);
+                let caller_balance = caller_account.balance;
+                state.account_read(
+                    &mut exec_step,
+                    call.caller_address,
+                    AccountField::Balance,
+                    caller_balance,
+                    caller_balance,
+                )?;
+            }
+            _ => (),
+        }
 
-        let (_, callee_account) = state.sdb.get_account(&call.address);
-        let is_empty_account = callee_account.is_empty();
-        let callee_nonce = callee_account.nonce;
-        state.account_read(
-            &mut exec_step,
-            call.address,
-            AccountField::Nonce,
-            callee_nonce,
-            callee_nonce,
-        )?;
         let callee_code_hash = call.code_hash;
-        let callee_code_hash_word = callee_code_hash.to_word();
-        state.account_read(
-            &mut exec_step,
-            callee_address,
-            AccountField::CodeHash,
-            callee_code_hash_word,
-            callee_code_hash_word,
-        )?;
+        let (callee_exists, _) = state.sdb.get_account(&callee_address);
```

Comment:
I think this might not be correct, the first bool is basically always true if the `StateDb` is constructed correctly, we might need to get `callee_exists` by checking that account indeed has some value.

```suggestion
        let callee_exists = !state.sdb.get_account(&callee_address).1.is_empty();
```